network/src/commonJvmAndroid/kotlin/com/wire/kalium/network/HttpEngine.kt

```diff
@@ -3,33 +3,42 @@ package com.wire.kalium.network
 import io.ktor.client.engine.HttpClientEngine
 import io.ktor.client.engine.okhttp.OkHttp
 import okhttp3.Interceptor
+import okhttp3.Request
 import okhttp3.Response
 
 actual fun defaultHttpEngine(): HttpClientEngine {
     return OkHttp.create {
-        /**
-         * This is a hack
-         *
-         * As seen here: https://github.com/ktorio/ktor/issues/1127
-         * Ktor is pretty has `application/protobuf` but not `application/x-protobuf`
-         * and it doesn't support custom content types!
-         */
         addInterceptor(object : Interceptor {
             override fun intercept(chain: Interceptor.Chain): Response {
-                val currentRequest = chain.call().request()
-                val currentHeaders = chain.call().request().headers
-                // When sending a byteArray, Ktor will put `octet-stream`. We can identify this and replace with protobuf
-                val isProtoBuf = currentHeaders["Content-Type"] == "application/octet-stream"
-
-                return if (!isProtoBuf) {
-                    chain.proceed(currentRequest)
-                } else {
-                    val newRequest = currentRequest.newBuilder()
-                        .header("Content-Type", "application/x-protobuf")
-                        .build()
-                    chain.proceed(newRequest)
-                }
+                return chain.call().request().handleProtoHeader()
+                    .let { checkedRequest -> chain.proceed(checkedRequest) }.handleUnauthorizedResponse()
             }
         })
     }
 }
+
+/**
+ * This is a hack
+ *
+ * As seen here: https://github.com/ktorio/ktor/issues/1127
+ * Ktor is pretty has `application/protobuf` but not `application/x-protobuf`
+ * and it doesn't support custom content types!
+ */
+private fun Request.handleProtoHeader(): Request =
+    when (headers["Content-Type"] == "application/octet-stream") {
+        true -> newBuilder().header("Content-Type", "application/x-protobuf").build()
+        false -> this
+    }
+
+/**
+ * Ktor need "WWW-Authenticate" to be set by BE in-order for the tokens refresh to work
+ * see issue https://youtrack.jetbrains.com/issue/KTOR-2806
+ * BE does not set "WWW-Authenticate"
+ *
+ * checks for 401 response -> add WWW-Authenticate header
+ */
+private fun Response.handleUnauthorizedResponse(): Response =
+    when (this.code == 401) {
+        true -> this.newBuilder().addHeader("WWW-Authenticate", "Bearer").build()
```

Comment:
Same than above

Reply:
Ktor need to know the type of auth method for the token refresh to work properly.
BE does not set the WWW-Authenticate, so this interceptor will detect 401 Unauthorized responses and add the needed header, I can do this a bit better and add the header only if it's missing.

Reply:
Maybe we can add some tests that contain other type of headers, and see if they survive and reach the backend after a token refresh interception?

Reply:
The Devs at square already done that for us 😅
https://github.com/square/okhttp/blob/master/okhttp/src/commonTest/kotlin/okhttp3/HeadersTest.kt